src/eckit/geo/Point3.h

```diff
@@ -0,0 +1,102 @@
+/*
+ * (C) Copyright 1996- ECMWF.
+ *
+ * This software is licensed under the terms of the Apache Licence Version 2.0
+ * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+ * In applying this licence, ECMWF does not waive the privileges and immunities
+ * granted to it by virtue of its status as an intergovernmental organisation nor
+ * does it submit to any jurisdiction.
+ */
+
+
+#pragma once
+
+#include <array>
+#include <ostream>
+
+
+namespace eckit::geo {
+
+
+class Point3 final : protected std::array<double, 3> {
+private:
+    // -- Types
+
+    using container_type = std::array<double, 3>;
+    using container_type::value_type;
+
+public:
+    // -- Constructors
+
+    Point3() :
+        Point3(0., 0., 0.) {}
+
+    Point3(double x, double y, double z) :
+        container_type{x, y, z} {}
+
+    Point3(const Point3& other) :
+        container_type(other) {}
+
+    Point3(Point3&& other) :
+        container_type(other) {}
+
+    // -- Destructor
+
+    ~Point3() = default;
+
+    // -- Operators
+
+    using container_type::operator[];
+
+    Point3& operator=(const Point3& other) {
+        container_type::operator=(other);
+        return *this;
+    }
+
+    Point3& operator=(Point3&& other) {
+        container_type::operator=(other);
+        return *this;
+    }
+
+    // -- Members
+
+    double& X = container_type::operator[](0);
+    double& Y = container_type::operator[](1);
+    double& Z = container_type::operator[](2);
```

Comment:
Following code should not be able to compile:
```c++
void my_function( const Point3& p ) {
  p.X = 1;
  p.Y = 2;
  p.Z = 3;
}
```
meaning that I should not be allowed to change the internal data on a `const Point3&`.
I'm not entirely sure that this is ensured.

Reply:
I've now refactored this taking into account the advice; There's an issue for const correctness because of brevity, but I've decided to enforce that points (all derived classes) are immutable to maintain internal consistency.